solr/core/src/java/org/apache/solr/core/ClusterEventProducerFactory.java

```diff
@@ -0,0 +1,178 @@
+  private Map<ClusterEvent.EventType, Set<ClusterEventListener>> initialListeners = new HashMap<>();
+  private CustomContainerPlugins.PluginRegistryListener initialPluginListener;
+  private final CoreContainer cc;
+  private boolean created = false;
+
+  public ClusterEventProducerFactory(CoreContainer cc) {
+    this.cc = cc;
+    initialPluginListener = new CustomContainerPlugins.PluginRegistryListener() {
+      @Override
+      public void added(CustomContainerPlugins.ApiInfo plugin) {
+        if (plugin == null || plugin.getInstance() == null) {
+          return;
+        }
+        Object instance = plugin.getInstance();
+        if (instance instanceof ClusterEventListener) {
+          registerListener((ClusterEventListener) instance);
+        }
+      }
+
+      @Override
+      public void deleted(CustomContainerPlugins.ApiInfo plugin) {
+        if (plugin == null || plugin.getInstance() == null) {
+          return;
+        }
+        Object instance = plugin.getInstance();
+        if (instance instanceof ClusterEventListener) {
+          unregisterListener((ClusterEventListener) instance);
+        }
+      }
+
+      @Override
+      public void modified(CustomContainerPlugins.ApiInfo old, CustomContainerPlugins.ApiInfo replacement) {
+        added(replacement);
+        deleted(old);
+      }
+    };
+  }
+
+  /**
+   * This method returns an initial plugin registry listener that helps to capture the
+   * freshly loaded listener plugins before the final cluster event producer is created.
+   * @return initial listener
+   */
+  public CustomContainerPlugins.PluginRegistryListener getPluginRegistryListener() {
+    return initialPluginListener;
+  }
+
+  /**
+   * Create a {@link ClusterEventProducer} based on the current plugin configurations.
+   * <p>NOTE: this method can only be called once because it has side-effects, such as
+   * transferring the initially collected listeners to the resulting producer's instance, and
+   * installing a {@link org.apache.solr.api.CustomContainerPlugins.PluginRegistryListener}.
+   * Calling this method more than once will result in an exception.</p>
+   * @param plugins current plugin configurations
+   * @return configured instance of cluster event producer (with side-effects, see above)
+   */
+  public ClusterEventProducer create(CustomContainerPlugins plugins) {
+    if (created) {
+      throw new RuntimeException("this factory can be called only once!");
+    }
+    final ClusterEventProducer clusterEventProducer;
+    CustomContainerPlugins.ApiInfo clusterEventProducerInfo = plugins.getPlugin(ClusterEventProducer.PLUGIN_NAME);
+    if (clusterEventProducerInfo != null) {
+      // the listener in ClusterSingletons already registered it
+      clusterEventProducer = (ClusterEventProducer) clusterEventProducerInfo.getInstance();
+    } else {
+      // create the default impl
+      clusterEventProducer = new DefaultClusterEventProducer(cc);
```

Comment:
There is no need to have a `ClusterEventProducer` if the user does not register it.

Reply:
I'm not sure. If we treat this as a part of core functionality (the ability to produce and listen to cluster-level events) then we need a rudimentary default implementation to be always present.

Reply:
This is a core , opt-in functionality

core = part of core
opt-in =. no traces if user does not explicitly add an event producer

Reply:
AFAIK there is no mechanism in plugins to require the presence of another plugin. If we decide to make this functionality an opt-in then other plugins that depend on this may silently fail to work properly because the user forgot to specify this opt-in. This problem is not specific to this plugin, but rather it's a lack of functionality in our whole plugin ecosystem.

Until we have a working mechanism to specify plugin requirements / dependencies I think we have to provide sensible defaults, and a sensible default is a working implementation, plus the ability to opt-out or change the default impl. I'll add a mechanism to opt-out.

Reply:
> AFAIK there is no mechanism in plugins to require the presence of another plugin.

There is no reason why we can't implement it. I think we should. Everything should be initialized as lazily as possible.

The add command for a `ClusterEventListener` has to do an extra check to see if a `ClusterEventProducer` is registered. If not, just register it

Reply:
> There is no reason why we can't implement it. I think we should.

Sure, but it's not a part of this PR. If/when we add it then we can drop this initialization of the default impl.

Reply:
NO , please. 

I am -1 on enabling this by default